### PR TITLE
[SPARK-14856] [SQL] returning batch correctly

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

Currently, the Parquet reader decide whether to return batch based on required schema or full schema, it's not consistent, this PR fix that.
## How was this patch tested?

Added regression tests.

### CODE REVIEW

**Comment by davies** [comment]
cc @sameeragarwal

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala [hunk 1]
```diff
@@ -308,8 +304,11 @@ private[sql] class DefaultSource
     // TODO: if you move this into the closure it reverts to the default values.
     // If true, enable using the custom RecordReader for parquet. This only works for
     // a subset of the types (no complex types).
-    val enableVectorizedParquetReader: Boolean = sqlContext.conf.parquetVectorizedReaderEnabled &&
-          dataSchema.forall(_.dataType.isInstanceOf[AtomicType])
+    val resultSchema = StructType(partitionSchema.fields ++ requiredSchema.fields)
+    val enableVectorizedReader: Boolean = sqlContext.conf.parquetVectorizedReaderEnabled &&
```
Comment (sameeragarwal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: minor: the comment above must be moved too

**Comment by sameeragarwal** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala (hunk 1 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: move to where?

Comment (sameeragarwal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: Nothing too important. The comment `// If true, enable using the custom RecordReader for parquet...` could be above this line.

**Comment by davies** [comment]
Merging this into master, thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala [hunk 2]
```diff
@@ -589,6 +590,30 @@ class ParquetQuerySuite extends QueryTest with ParquetTest with SharedSQLContext
...
+        assert(df2.queryExecution.sparkPlan.find(_.isInstanceOf[BatchedDataSourceScanExec]).isEmpty,
+          "Should not return batch")
+        checkAnswer(df2, df)
+
+        // return batch
+        val columns = Seq.tabulate(90) {i => s"c$i"}
+        val df3 = df2.selectExpr(columns : _*)
+        assert(
+          df3.queryExecution.sparkPlan.find(_.isInstanceOf[BatchedDataSourceScanExec]).isDefined,
+          "Should not return batch")
```
Comment (tedyu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala]: Should be "Should return batch"

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala [hunk 3]
```diff
@@ -589,6 +590,30 @@ class ParquetQuerySuite extends QueryTest with ParquetTest with SharedSQLContext
       checkAnswer(sqlContext.read.parquet(path), df)
     }
   }
+
+  test("returning batch for wide table") {
+    withSQLConf("spark.sql.codegen.maxFields" -> "100") {
```
Comment (tedyu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala]: Looks like the maxFields value can be set lower - resulting in faster test execution.

**Pull request closed without merging**